### PR TITLE
Machine Ownership

### PR DESCRIPTION
**Pull request by YoungOnionMC**
## What
Gives machines owners with UUIDs

## Implementation Details
any new machine placed will capture what placed it(just players for now) and set it as its owner

## Outcome
Potential future use for various things like wireless api, machine locking, etc

## Additional Information
could be expanded for compatibility for ftb teams

## Potential Compatibility Issues
any machine placed before this pr will have a null owner, but you can just break and replace it

### CODE REVIEW

File: src/main/java/com/gregtechceu/gtceu/api/blockentity/MetaMachineBlockEntity.java [hunk 1]
```diff
@@ -82,6 +90,15 @@ public MultiManagedStorage getRootStorage() {
         return managedStorage;
     }
 
+    @Override
+    public UUID getOwner() {
+        return owner;
+    }
+
+    public boolean ownerOnline() {
+        return ServerLifecycleHooks.getCurrentServer().getPlayerList().getPlayerByName(ownerName) != null;
```
Comment (screret) [inline src/main/java/com/gregtechceu/gtceu/api/blockentity/MetaMachineBlockEntity.java]: don't do this, get the owner by UUID instead.

File: src/main/java/com/gregtechceu/gtceu/common/blockentity/KineticMachineBlockEntity.java [hunk 2]
```diff
@@ -120,6 +121,11 @@ public MultiManagedStorage getRootStorage() {
         return managedStorage;
     }
 
+    @Override
+    public UUID getOwner() {
+        return null;
```
Comment (screret) [inline src/main/java/com/gregtechceu/gtceu/common/blockentity/KineticMachineBlockEntity.java]: implement the same things you did in MetaMachineBlockEntity here.

File: src/main/java/com/gregtechceu/gtceu/common/item/PortableScannerBehavior.java [hunk 3]
```diff
@@ -200,6 +201,9 @@ public int addScannerInfo(Player player, Level level, BlockPos pos, DisplayMode
 
             // General machine information
             if (mode == DisplayMode.SHOW_ALL || mode == DisplayMode.SHOW_MACHINE_INFO) {
+                if (machineBlockEntity instanceof MetaMachineBlockEntity mmBE)
+                    list.add(Component.translatable("behavior.portable_scanner.machine_ownership", mmBE.getOwnerName(),
```
Comment (screret) [inline src/main/java/com/gregtechceu/gtceu/common/item/PortableScannerBehavior.java]: why not just make these methods part of the IMachineBlockEntity interface?

**Review by screret — CHANGES_REQUESTED** [review]

**Review by Ghostipedia — CHANGES_REQUESTED** [review]
Make sure to natively support some kind of teams interface before this gets merged rather than doing it after.
Most common teams mod is FTBTeams & Argonauts, we should at a bear minimum find out how their team data is stored and also assign that as well if the player is in a team. 
Not having this in the initial API will result in addondevs not natively providing support either and cause more technical debt across the board than getting it done now.

File: src/main/java/com/gregtechceu/gtceu/api/blockentity/MetaMachineBlockEntity.java [hunk 4]
```diff
@@ -91,11 +89,28 @@ public MultiManagedStorage getRootStorage() {
...
+
     @Override
     public boolean ownerOnline() {
         return ServerLifecycleHooks.getCurrentServer().getPlayerList().getPlayer(owner) != null;
     }
 
+    @Override
+    public Class<?> getOwnerType() {
+        return ownerType;
+    }
```
Comment (screret) [inline src/main/java/com/gregtechceu/gtceu/api/blockentity/MetaMachineBlockEntity.java]: use `@Getter` for this

File: src/main/java/com/gregtechceu/gtceu/api/blockentity/MetaMachineBlockEntity.java [hunk 5]
```diff
@@ -91,11 +89,28 @@ public MultiManagedStorage getRootStorage() {
         return managedStorage;
     }
 
+    @Override
+    public UUID getOwnerUUID() {
+        return this.owner;
+    }
```
Comment (screret) [inline src/main/java/com/gregtechceu/gtceu/api/blockentity/MetaMachineBlockEntity.java]: use `@Getter` for this

File: src/main/java/com/gregtechceu/gtceu/common/blockentity/KineticMachineBlockEntity.java [hunk 6]
```diff
@@ -132,15 +131,27 @@ public MultiManagedStorage getRootStorage() {
...
+
     @Override
     public boolean ownerOnline() {
         return ServerLifecycleHooks.getCurrentServer().getPlayerList().getPlayer(owner) != null;
     }
 
+    @Override
+    public Class<?> getOwnerType() {
+        return ownerType;
+    }
```
Comment (screret) [inline src/main/java/com/gregtechceu/gtceu/common/blockentity/KineticMachineBlockEntity.java]: use `@Getter` for this

File: src/main/java/com/gregtechceu/gtceu/common/blockentity/KineticMachineBlockEntity.java [hunk 7]
```diff
@@ -132,15 +131,27 @@ public MultiManagedStorage getRootStorage() {
     }
 
     @Override
-    public UUID getOwner() {
+    public UUID getOwnerUUID() {
         return owner;
     }
```
Comment (screret) [inline src/main/java/com/gregtechceu/gtceu/common/blockentity/KineticMachineBlockEntity.java]: use `@Getter` for this

**Review by screret — CHANGES_REQUESTED** [review]
also add Argonauts support

**Comment by screret** [comment]
actually, this can't store the *team* as the owner anyway, since people can change teams. it should instead check if the using player and the owner are in the same team, I guess?

**Pull request closed without merging**